### DISCUSSION TITLE
How to programmatically parse/generate Prisma schema file?

### DISCUSSION THREAD

**Discussion by mbrain-io** [prisma/prisma#11006] [Q&A]
Is there a way to programmatically create/parse Prisma schema file?

**Comment by austincrim** [comment]
Hey @mbrain-io 👋 ,

Could you provide a little more context to your question? I'm not sure I totally understand what you are asking.

I will mention there is an undocumented `dmmf` attribute on the main `Prisma` export that contains a parsed representation of your Prisma schema.

```ts
import { Prisma } from '@prisma/client'

console.log(Prisma.dmmf)
```

Depending on what you want to accomplish, that might help.

**Reply by mbrain-io** [reply]
Thanx for the answer @austincrim 

I need to be able to let my users visually edit data structures. Visual editor saves its data in JSON format, so I need to programmatically create Prisma schema file from that JSON.

Also I need to programmatically read Prisma schema file and convert it back to JSON file so user can continue editing. I can not let users edit schema file directly, because users are not knowledgable enough to do that.

With dmmf second part is covered, but how do I programmatically create Prisma schema file?

**Reply by jonahallibone** [reply]
What DB are you using? This sounds like the wrong approach if you're using a relational db, and still sort of impossible using a predefined schema on a document db. I'd have a look at the [Json field type Prisma supports](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields), and see if that fits your needs better than creating a schema on the fly.

**Reply by mbrain-io** [reply]
@jonahallibone Thank you for the answer. Putting all custom fields in one JSON field sounds like a solution, but I dont know… Wouldn’t I loose many features? Type checking, relations between fields in JSON, etc. What do you think?

My goal is to let non programmers visually create schema, thats the project requirement. For example, similar how Strapi and many others let users define content types visually. It still looks to me that correct approach for my problem is to be able to programmatically create schema file, on the fly as you say it.

I am really surprised a bit, didn’t expect this to be an issue. I used Doctrine long time ago, and I think they let you write schema file in several standard formats (Yaml, XML). Can Prisma schema file be written in any other format for example?

**Comment by hyochan** [comment]
@mbrain-io Sounds like creating your own `Prisma Generator` can help your side.
https://github.com/YassinEldeeb/create-prisma-generator Could be the good initiative.

**Reply by mbrain-io** [reply]
Not sure I understand?

**Comment by Morriz** [comment]
We need the same, for the following use case:
1. prisma schema created
2. typegraphql-prisma integrated, which enables us to add decorators
3. generate classes with added decorators

**Comment by MitchellMonaghan** [comment]
Did you guy's create a solution for this? Would you mind sharing? Thanks

**Comment by assemmarwan** [comment]
Check [this](https://github.com/amplication/prisma-schema-dsl) out. It could be a solution for what you're looking for.

**Reply by mbrain-io** [reply]
@assemmarwan Thank you, will check it out!

**Comment by mohammed-bahumaish** [comment]
take a look at this repo, look into  `prisma-dmmf-extended` and `prisma-dmmf-modifier` packages , it may help.
https://github.com/mohammed-bahumaish/prisma-editor

**Comment by thim81** [comment]
hi @mbrain-io 

The snippet below will give you the full Prisma data modal from the Schema file.

```
import path, {dirname} from "path";
import {fileURLToPath} from "url";
import prismaInternals from '@prisma/internals';
const {getDMMF} = prismaInternals;

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

 const dmmf = await getDMMF({
      datamodelPath: path.join(__dirname, '../prisma/schema.prisma'),
    });
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
